Free Listings + Paid Ads: Add the ads audience and budget sections

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implements the **📌 Ads audience and budget sections** in #1610.

- Tweak UI for the `AudienceSection` component and `BudgetSection`-related components.
- Add the ads audience and budget sections to step 4 of the onboarding flow for paid campaign creation.
   - Handle the form data of campaign creation in `PaidAdsSectionsGroup` and forward it to the upper layer.
- Fix that the validation of campaign `amount` should only allow the number type (excluding `NaN`).

### Screenshots:

https://user-images.githubusercontent.com/17420811/186887919-b3fd78e4-2e62-4b26-88f3-9086758ecd15.mp4

### Detailed test instructions:

1. Go to step 4 of the onboarding flow.
1. Disconnect the Google Ads account via the current UI if an account is already connected.
1. Check if both the audience and budget sections are disabled when there is no connected Google Ads account.
1. Create or connect to a Google Ads account.
1. Check if both the audience and budget sections are editable after a Google Ads account is connected.
1. Check if the "Create a paid ad campaign" at the bottom is disabled when the audience countries are empty or the budget amount is ≤ 0.

💡 When re-entering step 4, the previously selected "continue setting up paid ads" state and entered form data are reset. Processing the selected state and entered form data will be developed in another PR.

💡 The disabled style covering the whole section is different from the visual on Figma. This's the expected style and has been confirmed with the Design team.

### Changelog entry

eason9487 committed:
Tweak UI for the AudienceSection component

eason9487 committed:
Tweak UI for the BudgetSection-related components

eason9487 committed:
Fix that the validation of campaign `amount` should only allow a number type (excluding NaN)

eason9487 committed:
Add the ads audience and budget sections to step 4 of the onboarding flow for paid campaign creation

eason9487 committed:
Use constants for the action values in the SetupPaidAds component.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1649#discussion_r957291065

eason9487 committed:
Two values were missed to be replaced with constants in the previous commit.

Ref: 3e0debdc862d734ea3d46ceb7782285474344ea2

puntope commented on js/src/setup-mc/setup-stepper/setup-paid-ads/setup-paid-ads.js:
```diff
@@ -19,19 +22,71 @@ import AppButton from '.~/components/app-button';
+			{ ( formProps ) => {
+				const { countryCodes } = formProps.values;
+				const disabledAudience =
+					googleAdsAccount?.status !== 'connected';
```
💅  This status could be also defined as Constant I guess

eason9487 replied:
Updated in https://github.com/woocommerce/google-listings-and-ads/pull/1655/commits/4d7da2b6bd972f1989b93c4cb23d6ec0d4fff171.

puntope commented on js/src/utils/paid-ads/__snapshots__/validateForm.test.js.snap:
```diff
@@ -0,0 +1,15 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`validateForm When the amount is not a number, should not pass 1`] = `"Please make sure daily average cost is greater than 0."`;
+
+exports[`validateForm When the amount is not a number, should not pass 2`] = `"Please make sure daily average cost is greater than 0."`;
+
+exports[`validateForm When the amount is not a number, should not pass 3`] = `"Please make sure daily average cost is greater than 0."`;
+
+exports[`validateForm When the amount is not a number, should not pass 4`] = `"Please make sure daily average cost is greater than 0."`;
+
+exports[`validateForm When the amount is ≤ 0, should not pass 1`] = `"Please make sure daily average cost is greater than 0."`;
+
+exports[`validateForm When the amount is ≤ 0, should not pass 2`] = `"Please make sure daily average cost is greater than 0."`;
+
+exports[`validateForm When the country codes array is empty, should not pass 1`] = `"Please select at least one country for your ads campaign."`;
```
💅  Maybe to consider switch to unit tests.

eason9487 replied:
Related discussions can be found in https://github.com/woocommerce/google-listings-and-ads/pull/893#discussion_r674040117 and https://github.com/woocommerce/google-listings-and-ads/pull/927#discussion_r681107540.

puntope commented on js/src/utils/paid-ads/validateForm.js:
```diff
@@ -25,7 +25,7 @@ const validateForm = ( values ) => {
 	}
 
-	if ( values.amount <= 0 ) {
+	if ( ! Number.isFinite( values.amount ) || values.amount <= 0 ) {
```
❓  How comes `amount` could be a non finite number?

eason9487 replied:
The budget amount will temporarily be set to `undefined` when the input is disabled.

https://github.com/woocommerce/google-listings-and-ads/blob/c08364ff753c5a2c50096e5b2254fd0c9ded0032/js/src/components/paid-ads/budget-section/index.js#L41-L50

puntope reviewed (approved):
✅  LGTM

- Tested the two sections match with figma and are disabled/enabled depending on if we have google ads connected. 

I left some comments in code (minor 💅 )

eason9487 committed:
Use constants for the status of Google Ads account.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1655#discussion_r960356644
